### ISSUE TITLE
Error on startup on a Mac

### ISSUE THREAD

**Issue by mutech** [lirantal/dockly#31]
## Expected Behavior

"UI" comes up

## Actual Behavior

HTML error page is displayed

## Steps to Re-produce the Problem

 1.
```
$ docker-machine env
export DOCKER_TLS_VERIFY="1"
export DOCKER_HOST="tcp://...:2376"
export DOCKER_CERT_PATH="/Users/.../.docker/machine/machines/default"
export DOCKER_MACHINE_NAME="default"
```
2.
```
$ npm install -g dockly
...
```
3.
```
$ dockly
Usage: dockly [OPTIONS]
...
Trace: { Error: (HTTP code 503) unexpected - <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 
...
			<div id="main">
				<h1>Websites are turned off.</h1>
				<h2>An administrator can turn them on using the Server application.</h2>
			</div>
...
    at /Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:254:17
    at getCause (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:284:7)
    at Modem.buildPayload (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:253:5)
    at IncomingMessage.<anonymous> (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:229:14)
    at emitNone (events.js:110:20)
    at IncomingMessage.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1045:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
  reason: undefined,
  statusCode: 503,
  json: '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"\n\t"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\n\n<html xmlns="http://www.w3.org/1999/xhtml" lang="en">\n\t<head>\n\t\t<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />\n\t\t<title>Server</title>\n\t\t<style type="text/css">\n\t\t* {\t\t\n\t\t\t-webkit-box-sizing: border-box;\n\t\t\t-moz-box-sizing: border-box;\n\t\t\t-box-sizing: border-box;\n\t\t}\n\t\tbody, html {\n\t\t\tbackground-color: #F2F3F4;\n\t\t\tcolor: #333333;\n\t\t\tfont-family: \'Lucida Grande\', \'Lucida Sans Unicode\', Helvetica, Arial, Verdana, sans-serif;\n\t\t\tfont-size: 14px;\n\t\t\theight: 100%;\n\t\t\tline-height: 21px;\n\t\t\tmargin: 0px;\n\t\t\ttext-align: center;\n\t\t\tword-spacing: -1px;\n\t\t}\n\t\t#content {\n\t\t\tpadding: 26px;\n\t\t}\n\t\t#main {\n\t\t\tbackground: #FFFFFF;\n\t\t\tborder: 1px solid #D5D5D6;\n\t\t\tborder-top-color: #E0E1E2;\n\t\t\tborder-bottom-color: #C0C1C2;\n\t\t\tmargin: 0px auto 0px auto;\n\t\t\tpadding: 100px 0px 100px 0px;\n\t\t\twidth: 730px;\n\t\t\t-webkit-box-shadow: 0px 1px 3px rgba(0,0,0,0.1);\n\t\t\t-moz-box-shadow: 0px 1px 3px rgba(0,0,0,0.1);\n\t\t\tbox-shadow: 0px 1px 3px rgba(0,0,0,0.1);\n\t\t\t-webkit-border-radius: 6px;\n\t\t\t-moz-border-radius: 6px;\n\t\t\tborder-radius: 6px;\n\t\t}\n\t\th1, h2 {\n\t\t\tcolor: #BBBBBB;\n\t\t\tfont-family: Helvetica, Arial, Verdana, sans-serif;\n\t\t\tfont-size: 18px;\n\t\t\tfont-weight: normal;\n\t\t\tline-height: 25px;\n\t\t\tmargin: 0px;\n\t\t\ttext-align: center;\n\t\t}\n\t\t</style>\n\t</head>\n\t<body>\n\t\t<div id="content">\n\t\t\t<div id="main">\n\t\t\t\t<h1>Websites are turned off.</h1>\n\t\t\t\t<h2>An administrator can turn them on using the Server application.</h2>\n\t\t\t</div>\n\t\t</div>\n\t</body>\n</html>\n' }
    at exitError (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:68:13)
    at initDockerConnection.then.then.catch (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:21:12)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
```
## Context

- Operating System: macOS high sierra
- Node.js version: 8.2.1
- Package version: latest (--version doesn't work either)
- Docker version: 17.09.0-ce
- Is docker installed locally? docker-machine with fusion driver
- Do you have containers created? yes
- Does this file exists `/var/run/docker.sock` ? no (see output of docker-machine env above)

**Comment by lirantal** [comment]
about the error message - you mean there's an http error message displayed back.

anyway though, are you absolutely sure that the DOCKER_HOST variable really points to the docker daemon? that message you got back doesn't seem to makes sense. It's an HTML response and you should be getting JSON responses from the docker daemon's API.

**Comment by lirantal** [comment]
@mutech I pushed a fix for an issue parsing the host when a DOCKER_HOST variable is provided.

Still, your value of it being `tcp://...:2376` is invalid. Just specify `host:port`

**Comment by mutech** [comment]
Hi @lirantal! Here is the full output of the command with ... replaced by actual values:

$ dockly
Usage: dockly [OPTIONS]
dockly [ --help | -v | --version ]

dockly

  Interactive shell for managing Docker containers

Options

  -u, --host string         Docker daemon hostname to connect to
  -p, --port string         Docker daemon port to connect to
  -s, --socketPath string   Docker daemon socket to connect to

  Project home: https://github.com/lirantal/dockly


Trace: { Error: (HTTP code 503) unexpected - <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
		<title>Server</title>
		<style type="text/css">
		* {
			-webkit-box-sizing: border-box;
			-moz-box-sizing: border-box;
			-box-sizing: border-box;
		}
		body, html {
			background-color: #F2F3F4;
			color: #333333;
			font-family: 'Lucida Grande', 'Lucida Sans Unicode', Helvetica, Arial, Verdana, sans-serif;
			font-size: 14px;
			height: 100%;
			line-height: 21px;
			margin: 0px;
			text-align: center;
			word-spacing: -1px;
		}
		#content {
			padding: 26px;
		}
		#main {
			background: #FFFFFF;
			border: 1px solid #D5D5D6;
			border-top-color: #E0E1E2;
			border-bottom-color: #C0C1C2;
			margin: 0px auto 0px auto;
			padding: 100px 0px 100px 0px;
			width: 730px;
			-webkit-box-shadow: 0px 1px 3px rgba(0,0,0,0.1);
			-moz-box-shadow: 0px 1px 3px rgba(0,0,0,0.1);
			box-shadow: 0px 1px 3px rgba(0,0,0,0.1);
			-webkit-border-radius: 6px;
			-moz-border-radius: 6px;
			border-radius: 6px;
		}
		h1, h2 {
			color: #BBBBBB;
			font-family: Helvetica, Arial, Verdana, sans-serif;
			font-size: 18px;
			font-weight: normal;
			line-height: 25px;
			margin: 0px;
			text-align: center;
		}
		</style>
	</head>
	<body>
		<div id="content">
			<div id="main">
				<h1>Websites are turned off.</h1>
				<h2>An administrator can turn them on using the Server application.</h2>
			</div>
		</div>
	</body>
</html>

    at /Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:254:17
    at getCause (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:284:7)
    at Modem.buildPayload (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:253:5)
    at IncomingMessage.<anonymous> (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/node_modules/docker-modem/lib/modem.js:229:14)
    at emitNone (events.js:110:20)
    at IncomingMessage.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1045:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
  reason: undefined,
  statusCode: 503,
  json: '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"\n\t"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\n\n<html xmlns="http://www.w3.org/1999/xhtml" lang="en">\n\t<head>\n\t\t<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />\n\t\t<title>Server</title>\n\t\t<style type="text/css">\n\t\t* {\t\t\n\t\t\t-webkit-box-sizing: border-box;\n\t\t\t-moz-box-sizing: border-box;\n\t\t\t-box-sizing: border-box;\n\t\t}\n\t\tbody, html {\n\t\t\tbackground-color: #F2F3F4;\n\t\t\tcolor: #333333;\n\t\t\tfont-family: \'Lucida Grande\', \'Lucida Sans Unicode\', Helvetica, Arial, Verdana, sans-serif;\n\t\t\tfont-size: 14px;\n\t\t\theight: 100%;\n\t\t\tline-height: 21px;\n\t\t\tmargin: 0px;\n\t\t\ttext-align: center;\n\t\t\tword-spacing: -1px;\n\t\t}\n\t\t#content {\n\t\t\tpadding: 26px;\n\t\t}\n\t\t#main {\n\t\t\tbackground: #FFFFFF;\n\t\t\tborder: 1px solid #D5D5D6;\n\t\t\tborder-top-color: #E0E1E2;\n\t\t\tborder-bottom-color: #C0C1C2;\n\t\t\tmargin: 0px auto 0px auto;\n\t\t\tpadding: 100px 0px 100px 0px;\n\t\t\twidth: 730px;\n\t\t\t-webkit-box-shadow: 0px 1px 3px rgba(0,0,0,0.1);\n\t\t\t-moz-box-shadow: 0px 1px 3px rgba(0,0,0,0.1);\n\t\t\tbox-shadow: 0px 1px 3px rgba(0,0,0,0.1);\n\t\t\t-webkit-border-radius: 6px;\n\t\t\t-moz-border-radius: 6px;\n\t\t\tborder-radius: 6px;\n\t\t}\n\t\th1, h2 {\n\t\t\tcolor: #BBBBBB;\n\t\t\tfont-family: Helvetica, Arial, Verdana, sans-serif;\n\t\t\tfont-size: 18px;\n\t\t\tfont-weight: normal;\n\t\t\tline-height: 25px;\n\t\t\tmargin: 0px;\n\t\t\ttext-align: center;\n\t\t}\n\t\t</style>\n\t</head>\n\t<body>\n\t\t<div id="content">\n\t\t\t<div id="main">\n\t\t\t\t<h1>Websites are turned off.</h1>\n\t\t\t\t<h2>An administrator can turn them on using the Server application.</h2>\n\t\t\t</div>\n\t\t</div>\n\t</body>\n</html>\n' }
    at exitError (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:68:13)
    at initDockerConnection.then.then.catch (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:21:12)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)

**Comment by mutech** [comment]
With version 2.3.1 I get this (no more HTML output):

$ dockly
Usage: dockly [OPTIONS]
dockly [ --help | -v | --version ]

dockly

  Interactive shell for managing Docker containers

Options

  -u, --host string         Docker daemon hostname to connect to
  -p, --port string         Docker daemon port to connect to
  -s, --socketPath string   Docker daemon socket to connect to

  Project home: https://github.com/lirantal/dockly


Trace: { Error: Parse Error
    at Socket.socketOnData (_http_client.js:454:20)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12)
    at readableAddChunk (_stream_readable.js:239:11)
    at Socket.Readable.push (_stream_readable.js:197:10)
    at TCP.onread (net.js:589:20) bytesParsed: 0, code: 'HPE_INVALID_CONSTANT' }
    at exitError (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:68:13)
    at initDockerConnection.then.then.catch (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:21:12)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)

$ docker-machine env
export DOCKER_TLS_VERIFY="1"
export DOCKER_HOST="tcp://172.16.145.142:2376"
export DOCKER_CERT_PATH="/Users/mu/.docker/machine/machines/default"
export DOCKER_MACHINE_NAME="default"

export DOCKER_HOST="172.16.145.142:2376" does not help (same result, makes sense since you apparently already got the connection)

With the same environment settings:

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                        PORTS               NAMES
043b5c242b85        mariadb:latest      "docker-entrypoint..."   8 days ago          Restarting (1) 1 second ago                       faircompany_mariadb_1

**Comment by lirantal** [comment]
can you explicitly specify https, such as:
```DOCKER_HOST="https://172.16.145.142:2376"```

**Comment by mutech** [comment]
using `DOCKER_HOST=https://172.16.145.145:2376` hung and finally failed with
```
Trace: { Error: connect ETIMEDOUT 172.16.145.145:2376
    at Object.exports._errnoException (util.js:1024:11)
    at exports._exceptionWithHostPort (util.js:1047:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1150:14)
  code: 'ETIMEDOUT',
  errno: 'ETIMEDOUT',
  syscall: 'connect',
  address: '172.16.145.145',
  port: 2376 }
    at exitError (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:68:13)
    at initDockerConnection.then.then.catch (/Users/mu/.nodenv/versions/8.2.1/lib/node_modules/dockly/index.js:21:12)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
```

Using certificates: Not consciously, though docker-machine seems to be configuring them. What I did was to install docker-machine (and related stuff) via home brew and then create the boot2docker image via docker-machine and the vmware fusion driver thereof.

I did not look into the docker API yet, but I assume if the node lib you're using is not reading the DOCKER_* env. variables you might need to explicitely enable TLS with the DOCKER_CERT_PATH provided or maybe TLS is not even supported.

**Comment by mutech** [comment]
If you can't reproduce and want to look into it, we could meet up via teamviewer or something

**Comment by lirantal** [comment]
Right. So the lib I'm using supports the following:
```
    docker = new Docker({
        host: '192.168.99.100',
        protocol: 'https',
        ca: fs.readFileSync(BASE_PATH + 'ca.pem'),
        cert: fs.readFileSync(BASE_PATH + 'cert.pem'),
        key: fs.readFileSync(BASE_PATH + 'key.pem'),
        port: 2376
    });
```

Which out of that I'm only passing on the `host` and `port` parameters.
I can expose env vars for each of the cert parameters (key, cert, ca). Will that work for you?

I found this should be supported on Windows if you want to try:
`DOCKER_SOCKET=//./pipe/docker_engine`

**Comment by mutech** [comment]
I think it should be fine to set BASE_PATH in your example to the content of $DOCKER_CERT_PATH and set properties ca, cert and key like you did; if DOCKER_CERT_PATH is defined.

The referenced directory on my box contains the files ca.pem, cert.pem and key.pem, so if the certificate is the reason for the failure we should be good.

**Comment by lirantal** [comment]
That BASE_PATH is unrelated, it's just the directory path to find the files, but yeah like you said if you have those files I can make a change to take them into account.

It's a quick one, I'll work on a fix for that and ping you in an hour.

**Comment by mutech** [comment]
Did just that (hacky):

```
const DOCKER_HOST = process.env.DOCKER_HOST || false
const DOCKER_CERT_PATH = process.env.DOCKER_CERT_PATH || false

let dockerCon

function util (connection) {
  if (typeof connection !== 'object') {
    throw new Error('Error: docker connection string is faulty, please review command line arguments.')
  }

  if (!connection.hasOwnProperty('socketPath') && !connection.hasOwnProperty('host')) {
    // attempt honoring the DOCKER_HOST variable, and fallback to connect using the
    // the docker daemon socket
    if (DOCKER_HOST) {
      const parsedUrl = parseUrl(DOCKER_HOST)
      if (DOCKER_CERT_PATH) {
        // TODO: use fs module to handle paths
        const certPath = DOCKER_CERT_PATH;
        dockerCon = new DockerLib({host: parsedUrl.hostname, port: parsedUrl.port, ca: certPath + '/ca.pem', cert: certPath + 'cert.pem', key: certPath + 'key.pem'})
      } else {
        dockerCon = new DockerLib({host: parsedUrl.hostname, port: parsedUrl.port})
      }
```

but that just leads to the next problem:

```
Trace: Error: error:0906D06C:PEM routines:PEM_read_bio:no start line
    at Object.createSecureContext (_tls_common.js:89:17)
    at Object.exports.connect (_tls_wrap.js:1038:48)
    at Agent.createConnection (https.js:111:22)
    at Agent.createSocket (_http_agent.js:214:26)
    at Agent.addRequest (_http_agent.js:184:10)
    at new ClientRequest (_http_client.js:272:16)
    at Object.request (http.js:39:10)
    at Object.request (https.js:239:15)
    at h.request (/Users/mu/git/_3rd party/dockly/node_modules/docker-modem/lib/http.js:51:52)
    at Modem.buildRequest (/Users/mu/git/_3rd party/dockly/node_modules/docker-modem/lib/modem.js:193:33)
    at exitError (/Users/mu/git/_3rd party/dockly/index.js:68:13)
    at initDockerConnection.then.then.catch (/Users/mu/git/_3rd party/dockly/index.js:21:12)
    at <anonymous>
    at runMicrotasksCallback (internal/process/next_tick.js:121:5)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
    at Function.Module.runMain (module.js:607:11)
    at startup (bootstrap_node.js:158:16)
    at bootstrap_node.js:575:3
```

**Comment by lirantal** [comment]
thanks for jumping on this and making the change locally, definitely speed things up.
I found this if it helps: https://stackoverflow.com/questions/38703443/keep-getting-error0906d06cpem-routinespem-read-biono-start-line

**Comment by mutech** [comment]
Could do, but it would be ugly, since I am (users of docker-machine are not) directly involved in maintaining the certificates. Since all other docker tools get along well with this, it would be a bit painful to have to do that manually.

**Comment by mutech** [comment]
I'm silly, overlooked that the API wants the data and not the path/url to cert files. Here is the code fragment that works for me:

```
const DOCKER_CERT_PATH = process.env.DOCKER_CERT_PATH || false

let dockerCon

function util (connection) {
  if (typeof connection !== 'object') {
    throw new Error('Error: docker connection string is faulty, please review command line arguments.')
  }

  if (!connection.hasOwnProperty('socketPath') && !connection.hasOwnProperty('host')) {
    // attempt honoring the DOCKER_HOST variable, and fallback to connect using the
    // the docker daemon socket
    if (DOCKER_HOST) {
      const parsedUrl = parseUrl(DOCKER_HOST)
      let parameters = {host: parsedUrl.hostname, port: parsedUrl.port};

      if (DOCKER_CERT_PATH) {
        const base = new url.URL(DOCKER_CERT_PATH.endsWith('/') ? DOCKER_CERT_PATH : DOCKER_CERT_PATH + '/', 'file://');
        ['ca', 'cert', 'key'].forEach(
          function(fileName) {
            const fileUrl = new url.URL(fileName + '.pem', base)
            const data = fs.readFileSync(fileUrl)
            if (data) {
              parameters[fileName] = data
            }
          })
        dockerCon = new DockerLib(parameters)

      } else {
        dockerCon = new DockerLib({socketPath: DOCKER_SOCKET})
      }
    } else {
      dockerCon = new DockerLib(connection)
    }
  }
}
```

**Comment by lirantal** [comment]
Oh good catch!
I'm so happy you were able to work it out :-)

Would you like to open a PR with those changes? I'll be happy to merge and release a new version with support for SSL.

**Comment by mutech** [comment]
Can do it tomorrow, though I guess it's faster if you just copy over the couple of lines above if you feel they're ok as is ;)

**Comment by lirantal** [comment]
Sure, no problem. I'll push a new version tomorrow and ping you, and if you won't open a PR yourself by tomorrow I'll do it, no worries.

Thanks a lot for being patient but mostly for pursuing it and working it out!

![](https://media.giphy.com/media/xT0BKhunZXlEsnpz7q/giphy-downsized.gif)

**Comment by lirantal** [comment]
@mutech I removed all that connection handling from dockly and offloaded it to the underlying docker library I use (dockerode) so it should already handle all of these environment variables and such.

I released dockly 3.1.0 if you care to upgrade and test (and use :-) )
